Framework/API/src/FunctionProperty.cpp

```diff
@@ -16,20 +16,24 @@ namespace API {
  *  Sets the property names but initialises the function pointer to null.
  *  @param name :: The name to assign to the property
  *  @param direction :: The direction of the function (i.e. input or output)
+ *  @param optional :: A flag indicating whether the property is optional or mandatory.
  */
-FunctionProperty::FunctionProperty(const std::string &name, const unsigned int direction)
+FunctionProperty::FunctionProperty(const std::string &name, const unsigned int direction,
+                                   const PropertyMode::Type optional)
     : Kernel::PropertyWithValue<std::shared_ptr<IFunction>>(
-          name, std::shared_ptr<IFunction>(), Kernel::IValidator_sptr(new Kernel::NullValidator()), direction) {}
+          name, std::shared_ptr<IFunction>(), Kernel::IValidator_sptr(new Kernel::NullValidator()), direction),
+      m_optional(optional) {}
 
 /// Copy constructor
 FunctionProperty::FunctionProperty(const FunctionProperty &right)
-    : Kernel::PropertyWithValue<std::shared_ptr<IFunction>>(right) {}
+    : Kernel::PropertyWithValue<std::shared_ptr<IFunction>>(right), m_optional(right.m_optional) {}
 
 /// Copy assignment operator. Copies the pointer to the function.
 FunctionProperty &FunctionProperty::operator=(const FunctionProperty &right) {
-  if (&right == this)
+  if (&right == this && right.m_optional == m_optional)
     return *this;
   Kernel::PropertyWithValue<std::shared_ptr<IFunction>>::operator=(right);
+  m_optional = right.m_optional;
   return *this;
 }
```

Comment:
These two can be removed rather than adding to them. The compiler knows how to generate defaults for both (as a std:: type and a custom type we've implemented a copy const. for) and writing copy constructors is fraught with bugs and perils so always makes me nervous.

Instead can we delete this entire block for the copy assignment and constructor operators and in the header simply append `= default;` to both. This way we know 100% we haven't missed something out on our copy constructor